### ISSUE TITLE
[Release-1.21] - starting kubernetes: preparing server: json: cannot unmarshal string into Go value of type bootstrap.File

### ISSUE THREAD

**Issue by briandowns** [k3s-io/k3s#4731]
Backport fix for starting kubernetes: preparing server: json: cannot unmarshal string into Go value of type bootstrap.File

* #4644

**Comment by ShylajaDevadiga** [comment]
To reproduce the issue,

- created a 2 server, 1 worker node cluster with v1.21.5+k3s2
- Tried to join 3rd node with v1.21.7+k3s1

Node failed to join the cluster with error mentioned above
```
Dec 15 03:52:43 ip-172-31-7-32.us-east-2.compute.internal k3s[6844]: time="2021-12-15T03:52:43.960777643Z" level=fatal msg="starting kubernetes: preparing server: json: cannot unmarshal string into Go value of type bootstrap.File"

$ kubectl get nodes
NAME                                         STATUS   ROLES                       AGE    VERSION
ip-172-31-3-168.us-east-2.compute.internal   Ready    control-plane,etcd,master   11m    v1.21.5+k3s2
ip-172-31-4-235.us-east-2.compute.internal   Ready    control-plane,etcd,master   5m2s   v1.21.5+k3s2
ip-172-31-5-1.us-east-2.compute.internal     Ready    <none>                      14s    v1.21.5+k3s2
```
Validated the fix using **v1.21.7-rc2+k3s2 **
To the above cluster joined the node using version v1.21.7-rc2+k3s2 successfully
```
$ kubectl get nodes 
NAME                                         STATUS   ROLES                       AGE   VERSION
ip-172-31-3-168.us-east-2.compute.internal   Ready    control-plane,etcd,master   30m   v1.21.5+k3s2
ip-172-31-4-235.us-east-2.compute.internal   Ready    control-plane,etcd,master   24m   v1.21.5+k3s2
ip-172-31-5-1.us-east-2.compute.internal     Ready    <none>                      19m   v1.21.5+k3s2
ip-172-31-7-32.us-east-2.compute.internal    Ready    control-plane,etcd,master   96s   v1.21.7-rc2+k3s2
```
Repeated on a 2 node setup
Node1:
```
curl -sfL https://get.k3s.io |INSTALL_K3S_VERSION=v1.21.5+k3s2  INSTALL_K3S_TYPE='server' sh -s - server --cluster-init --token <token>
```
Node2
```
curl -sfL https://get.k3s.io |INSTALL_K3S_VERSION=v1.21.7-rc2+k3s2  INSTALL_K3S_TYPE='server' sh -s - server --server https://<server1 ip>:6443 --token <token>
```
Node 2 joined successfully
```
$ kubectl get nodes 
NAME                                          STATUS   ROLES                       AGE   VERSION
ip-172-31-13-136.us-east-2.compute.internal   Ready    control-plane,etcd,master   11m   v1.21.7-rc2+k3s2
ip-172-31-14-229.us-east-2.compute.internal   Ready    control-plane,etcd,master   14m   v1.21.5+k3s2

$ kubectl get pods -A
NAMESPACE     NAME                                      READY   STATUS      RESTARTS   AGE
kube-system   coredns-7448499f4d-5wgk8                  1/1     Running     0          14m
kube-system   helm-install-traefik-crd-qr4zq            0/1     Completed   0          11m
kube-system   helm-install-traefik-xk9bz                0/1     Completed   1          11m
kube-system   local-path-provisioner-5ff76fc89d-hrkt9   1/1     Running     0          14m
kube-system   metrics-server-86cbb8457f-gb2nh           1/1     Running     0          14m
kube-system   svclb-traefik-m6q9k                       2/2     Running     0          14m
kube-system   svclb-traefik-qsv8j                       2/2     Running     0          11m
kube-system   traefik-6b84f7cbc-ks27t                   1/1     Running     0          11m
```
Tried to reproduce upgrade scenario using the following setup. Issue was not reproduced. Upgrade was successful.

- Created a 3 node cluster with v1.21.5+k3s2
- Skipped upgrading node1. Started with upgrading node2 to v1.21.7+k3s1. Upgrade was successful
- Similarly upgraded node3 as well as node1 without issues to v1.21.7+k3s1

Note: In all scenarios token was passed in every node

**Comment by ShylajaDevadiga** [comment]
Closing as the issue seems to be resolved as mentioned here #4644 (comment)
